[Translation] Allow default parameters

Jean-Beru opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 7.3
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no 
| Issues        | Fix #48182
| License       | MIT

This PRs adds a `GlobalsTranslator` which decorates the `Translator` when global parameters are given.

Usage:
```yaml
# config/framework.yaml
framework:
    # ...
    translator:
        # ...
        globals:
            '{version}': '1.2.3'
            '%%application_name%%': 'My app' # "%" must be escaped if it is not a DI parameter
```
 
```yaml
# messages.en.yaml
app:
  version: 'Application version: {version}'
  name: 'Application name: %application_name%'
```

```twig
# twig template
{{ 'app.version'|trans }} # Displays "Application version: 1.2.3"
{{ 'app.version'|trans({ '{version}': '2.3.4' }) }} # Displays "Application version: 2.3.4"
{{ 'app.name'|trans }} # Displays "Application name: My app"
{{ 'app.name'|trans({ '{application_name}': 'My new app' }) }} # Displays "Application name: My new app"
```

Profiler view:
![image](https://github.com/user-attachments/assets/6ba13995-ca91-4098-a87d-daea89864ab6)

Jean-Beru committed:
[Translation] Allow default parameters

Jean-Beru commented:
Friendly ping @94noni :wink:

94noni commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -1580,6 +1581,18 @@ private function registerTranslatorConfiguration(array $config, ContainerBuilder
 
+        if (!empty($config['globals'])) {
+            $def = $container
+                ->register('translator.globals', GlobalsTranslator::class)
```
what about using the default `translator` service? 
I mean for twig, it is directly inside the `twig` service

Jean-Beru replied:
It was my first idea. What I had in mind:
* Create a new interface (in Contract ?)
* Tag services with this interface
* Call `addGlobal` with parameters given in configuration
It seems a little bit more complex.

94noni reviewed (commented):
nice :)
just a small comment regarding if we can add this feature on the default translator or not
obviously in our app we defined a decorator, but here it is core/new feature so I dont see why a new class is needed (or did I miss something?)

(and your example in PR desc part "# twig template" seems wrong for default using global)

### wrong

```twig
{{ 'app.version'|trans }} # Displays "1.2.3"
{{ 'app.version'|trans({ '{version}': '2.3.4' }) }} # Displays "2.3.4"
```

### correct

```twig
{{ 'app.version'|trans }} # Displays "Application version: 1.2.3"
{{ 'app.version'|trans({ '{version}': '2.3.4' }) }} # Displays "Application version: 2.3.4"
```

94noni commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php:
```diff
@@ -983,6 +984,13 @@ private function addTranslatorSection(ArrayNodeDefinition $rootNode, callable $e
                         ->end()
+                        ->arrayNode('globals')
+                            ->info('Global parameters.')
+                            ->example(['pi' => 3.14])
```
I would use the '`{app_version}`' example as it also showcase the delimiters

Jean-Beru replied:
Good idea, it's done.

94noni reviewed (approved):
lgtm, lets wait other reviews :)

nicolas-grekas commented on src/Symfony/Component/Translation/GlobalsTranslator.php:
```diff
@@ -0,0 +1,41 @@
+
+use Symfony\Contracts\Translation\TranslatorInterface;
+
+final class GlobalsTranslator implements TranslatorInterface
```
I would call that "Defaults", not globals (the scope is local in the implementation)

Jean-Beru replied:
I renamed it to `DefaultParameters` to distinguish from `translator.default` service.

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php:
```diff
@@ -929,6 +929,7 @@ private function addTranslatorSection(ArrayNodeDefinition $rootNode, callable $e
                     ->fixXmlConfig('fallback')
                     ->fixXmlConfig('path')
                     ->fixXmlConfig('provider')
+                    ->fixXmlConfig('default_parameter')
```
or just "parameter"?

welcoMattic replied:
Or "global_parameters"?

nicolas-grekas replied:
or just "global" like in twig? :D

Jean-Beru replied:
So let's go for the following naming ?

`globals` in configuration
```yaml
framework:
    translator:
        globals:
            '{version}': '1.2.3'
            '{project_dir}': '%kernel.project_dir%'
```

And `default_parameters` for the service
```php
$container->register('translator.default_parameters', DefaultParametersTranslator::class)
```

nicolas-grekas replied:
yes please :)

Jean-Beru replied:
Done :slightly_smiling_face:

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -1580,6 +1581,18 @@ private function registerTranslatorConfiguration(array $config, ContainerBuilder
             }
         }
 
+        if (!empty($config['default_parameters'])) {
```
we try to never use empty in the codebase

Jean-Beru replied:
OK. Changed for a `[] === ` comparison.

nicolas-grekas commented on src/Symfony/Component/Translation/DefaultParametersTranslator.php:
```diff
@@ -0,0 +1,41 @@
+        return $this->translator->getLocale();
+    }
+
+    public function addDefaultParameter(string $name, mixed $value): void
```
it'd be better to replace this method by a constructor argument. would that work for you?

Jean-Beru replied:
No problem, it's done.

welcoMattic commented on src/Symfony/Component/Translation/Tests/DefaultParametersTranslatorTest.php:
```diff
@@ -0,0 +1,41 @@
+        $this->assertSame('Welcome Global name!', $defaultParametersTranslator->trans('test'));
+        $this->assertSame('Welcome Name!', $defaultParametersTranslator->trans('test', ['%name%' => 'Name']));
+    }
+
```
Could you add more tests to cover both Message format and Intl format please?

Jean-Beru replied:
I added some tests with a different locale and with the ICU format. Is it OK for you ?

Jean-Beru commented:
Could it be considered as a BC break since we can use a public method from the `Translator` service (from implemented interfaces or not) but not present in this decorator?

luxferre commented:
Can a global parameter value reference translation key?

Jean-Beru commented:
> Can a global parameter value reference translation key?

For now, in this PR, a global parameter can be used with a fixed value, a parameter or an environment variable.

stof commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php:
```diff
@@ -985,6 +986,14 @@ private function addTranslatorSection(ArrayNodeDefinition $rootNode, callable $e
+                        ->arrayNode('globals')
+                            ->info('Global parameters.')
+                            ->example(['app_version' => 3.14])
+                            ->defaultValue([])
```
this is useless as it is already the default behavior for prototypes array nodes.

Jean-Beru replied:
Good catch

stof commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -1567,6 +1568,16 @@ private function registerTranslatorConfiguration(array $config, ContainerBuilder
+        if ([] !== $config['globals']) {
+            $container
+                ->register('translator.default_parameters', DefaultParametersTranslator::class)
+                ->setDecoratedService('translator', null, -1) // Lower priority than "translator.data_collector"
```
do we need a specific priority compared to the PseudoLocalizationTranslator ?

Jean-Beru replied:
This priority is needed to avoid the following error: `The Translator "Symfony\Component\Translation\DefaultParametersTranslator" must implement TranslatorInterface, TranslatorBagInterface and LocaleAwareInterface.`.

On reflection, since the `DefaultParametersTranslator` only implements `TranslatorInterface`, using it could limit usages of the initial service (ex: I want to use the `getCatalogues` method from TranslatorBagInterface`).

@stof Do you think that it will be preferable to add this behavior to the `Translator` like mentioned in https://github.com/symfony/symfony/pull/53425#discussion_r1442025289 ?

welcoMattic replied:
IIUC the priority meaning, I guess we should put `DefaultParametersTranslator` before `PseudoLocalizationTranslator`, in order to leverage the pseudo-localization for replaced values of defaults parameters. WDYT?

Jean-Beru replied:
> IIUC the priority meaning, I guess we should put `DefaultParametersTranslator` before `PseudoLocalizationTranslator`, in order to leverage the pseudo-localization for replaced values of defaults parameters. WDYT?

Not really because `PseudoLocalizationTranslator` calls the decorated stack before updating the translated text. Also it does not rely on parameters, which are the only thing changed by `DefaultParametersTranslator`.

See https://github.com/symfony/symfony/blob/9519b54417c09c49496a4a6be238e63be9a73465/src/Symfony/Component/Translation/PseudoLocalizationTranslator.php#L91

stof commented on src/Symfony/Bundle/FrameworkBundle/Resources/config/schema/symfony-1.0.xsd:
```diff
@@ -231,13 +231,15 @@
         <xsd:attribute name="formatter" type="xsd:string" />
         <xsd:attribute name="cache-dir" type="xsd:string" />
         <xsd:attribute name="default-path" type="xsd:string" />
+        <xsd:attribute name="global" type="xsd:string" />
```
this attribute does not make sense.

Jean-Beru replied:
Indeed!

94noni reviewed (approved):
Nice :)

welcoMattic commented on src/Symfony/Component/Translation/CHANGELOG.md:
```diff
@@ -5,6 +5,7 @@ CHANGELOG
 ---
 
  * Mark class `DataCollectorTranslator` as `final`
+ * Add `DefaultParametersTranslator` to allow defining default parameters
```
```suggestion
 * Add `DefaultParametersTranslator` to allow defining default translation parameters
```

welcoMattic commented on src/Symfony/Component/Translation/Tests/DefaultParametersTranslatorTest.php:
```diff
@@ -0,0 +1,64 @@
+        $this->assertSame('Bienvenue Jean!', $globalsTranslator->trans('welcome', ['{name}' => 'Jean'], null, 'fr'));
+    }
+
+    public function testTransICU()
+    {
+        if (!class_exists(\MessageFormatter::class)) {
+            $this->markTestSkipped(sprintf('Skipping test as the required "%s" class does not exist. Consider installing the "intl" PHP extension or the "symfony/polyfill-intl-messageformatter" package.', \MessageFormatter::class));
+        }
```
```suggestion
    /**
     * @requires extension intl
     */
    public function testTransICU()
    {
```

Jean-Beru replied:
Is this class necessarily loaded? If so, should we will have to update https://github.com/symfony/symfony/blob/4ad506b0855f45061c2d4b2600846a8aaf338f98/src/Symfony/Component/Translation/Tests/TranslatorCacheTest.php#L62 and https://github.com/symfony/symfony/blob/33d71aa7b7c368e9752198413c9fbd0ca8d82617/src/Symfony/Component/Translation/Tests/TranslatorTest.php#L406 too.

welcoMattic replied:
Yes, this class lives in the intl extension. So, marking the test requiring the extension avoid to check on the class existence on all tests that require intl extension. Am I wrong @xabbuh?

Jean-Beru replied:
Indeed. I updated the tests this way.

nicolas-grekas reviewed (commented):
I'm wondering if we really need a decorator: adding this concern on the existing Translator looks trivial to me.

94noni commented:
[I totally agree](https://github.com/symfony/symfony/pull/53425#discussion_r1442025289)
would love to have it on main service :)

Jean-Beru commented:
@nicolas-grekas @94noni Indeed, much simpler like that :+1: 

I added the global parameters list to the profiler to make debugging easier. A screenshot is available in the description of this PR.

94noni commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/translation.html.twig:
```diff
@@ -156,6 +156,27 @@
+            {% for name, value in collector.globalParameters %}
+                <tr>
+                    <td class="font-normal text-small text-bold nowrap">{{ name }}</td>
+                    <td class="font-normal text-small nowrap">{{ value }}</td>
```
as globalParameters is an array of "string, mixed"
how will `{{ value }}` be rendered here?
perhaps use the `profiler_dump()` function?

also while talking about mixed, perhaps add an example / test / doc of for example a global array params ? (if handled as is)

```yaml
globals:
    hosts:
        'FR': 'https://foo.bar.fr'
        'ES': 'https://foo.bar.es'
```

// totaly arbitrary example

Jean-Beru replied:
For now configuration only allows scalar values. To make it simpler we could restrict to `int|float|string| for now.

Your example seems to be related to [this comment](https://github.com/symfony/symfony/pull/53425#issuecomment-1974769159). Since a parameter can be a `TranslatableMessage`, it could be interesting to allow an other way to configure global parameters:
```yaml
# Use a value
globals:
    host: 
        value: 'https://unique.site.com' # use this value

# Use a value (short alias)
globals:
    host: 'https://unique.site.com' # use this value

# Use a message key
globals:
    host:
        message: 'host' # use TranslatableMessage('host') with the same parameters and domain

# Use a message key from an other domain
globals:
    host:
        message: 'host' # use TranslatableMessage('host') with the same parameters but "common" domain
        domain: 'common'
```

WDYT ?

94noni replied:
Yes can makes sens

Jean-Beru commented:
I had some difficulties with infinite loops but `TranslatableMessage` can be used as globals now. :tada:

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php:
```diff
@@ -1020,6 +1021,33 @@ private function addTranslatorSection(ArrayNodeDefinition $rootNode, callable $e
+                                ->end()
+                                ->beforeNormalization()
+                                    ->ifTrue(fn ($v) => !\is_array($v))
+                                    ->then(fn ($v): array => ['value' => $v])
```
the closures could all be made static also
```suggestion
                                    ->then(fn ($v) => ['value' => $v])
```

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -1594,6 +1594,16 @@ private function registerTranslatorConfiguration(array $config, ContainerBuilder
             $translator->replaceArgument(4, $options);
         }
 
+        if ([] !== $config['globals']) {
```
```suggestion
        if ($config['globals']) {
```

nicolas-grekas replied:
but actually there is no need for this "if" nesting at all

nicolas-grekas commented on src/Symfony/Component/Translation/Translator.php:
```diff
@@ -174,6 +210,13 @@ public function trans(?string $id, array $parameters = [], ?string $domain = nul
             }
         }
 
+        if ([] !== $this->globalTranslatableParameters && !isset($this->globalParameters[$locale])) {
```
we don't use empty-values comparisons in Symfony (same everywhere and for your next PRs ;) )
```suggestion
        if ($this->globalTranslatableParameters && !isset($this->globalParameters[$locale])) {
```

Jean-Beru replied:
Noted, thanks :slightly_smiling_face:

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/Fixtures/php/translator_globals.php:
```diff
@@ -0,0 +1,15 @@
+        'globals' => [
+            '%%app_name%%' => 'My application',
+            '{app_version}' => '1.2.3',
+            'url' => ['message' => 'url', 'parameters' => ['scheme' => 'https://'], 'domain' => 'global'],
```
even in tests, the key should use delimiters to be clear about the fact it's going to be needed in real apps also

welcoMattic commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/translation.html.twig:
```diff
@@ -156,6 +156,31 @@
+            {% for locale, values in collector.globalParameters %}
+                {% for id, value in values %}
+                    <tr>
+                        <td class="font-normal text-small nowrap">{{ locale != '*' ? "#{locale}" }}</td>
```
```suggestion
                        <td class="font-normal text-small nowrap">{{ locale != '*' ? "{locale}" }}</td>
```

To stay consistent with existing table in translation panel
<img width="404" alt="image" src="https://github.com/user-attachments/assets/2e364e04-7cf6-4cff-aad9-77e3c21c835f" />

Jean-Beru replied:
I'm not sure to understand. If we remove `#`, the string interpolation won't be done and it will lead to a column with `{locale}` values.

stof replied:
why using a string containing *only an interpolation of a variable ? Use the variable directly instead.
In PHP, are you using `"$myVar"` instead of `$myVar` when you want to pass a string coming from the `$myVar` variable to another part of the code ?

Jean-Beru replied:
I wanted to go too fast :facepalm:

It's fixed.

welcoMattic commented on src/Symfony/Component/Translation/Translator.php:
```diff
@@ -155,6 +165,32 @@ public function getFallbackLocales(): array
         return $this->fallbackLocales;
     }
 
+    /**
+     * All a global parameter.
+     */
```
Not necessary, I think the method name is self-describing

welcoMattic commented on src/Symfony/Component/Translation/Translator.php:
```diff
@@ -155,6 +165,32 @@ public function getFallbackLocales(): array
+        $this->globalParameters['*'][$id] = $value;
+    }
+
+    /**
+     * All a global translatable parameter.
+     */
```
Not necessary, I think the method name is self-describing

welcoMattic commented on src/Symfony/Component/Translation/Translator.php:
```diff
@@ -155,6 +165,32 @@ public function getFallbackLocales(): array
+    }
+
+    /**
+     * Gets the global parameters.
```
Not necessary, I think the method name is self-describing

welcoMattic reviewed (commented):
Minor comments, otherwise it's ✅ for me!
@Jean-Beru could you squash your commits please?

Jean-Beru commented:
> Minor comments, otherwise it's ✅ for me! @Jean-Beru could you squash your commits please?

Thanks @welcoMattic !

PR is fixed, squashed and rebased.

stof commented on src/Symfony/Bundle/FrameworkBundle/Resources/config/schema/symfony-1.0.xsd:
```diff
@@ -285,6 +286,24 @@
+        <xsd:attribute name="domain" type="xsd:string" />
+    </xsd:complexType>
+
+    <xsd:complexType name="translation_global_parameter" mixed="true">
```
why `mixed="true` ?

stof commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php:
```diff
@@ -1020,6 +1021,33 @@ private function addTranslatorSection(ArrayNodeDefinition $rootNode, callable $e
+                            ->arrayPrototype()
+                                ->fixXmlConfig('parameter')
+                                ->children()
+                                    ->variableNode('value')->end()
```
This should be a scalar node, not a variable node. There is no reason to support non-scalar values as those would fail at runtime (an array cannot be casted to string to be inserted in the translation)

stof commented on src/Symfony/Component/Translation/Translator.php:
```diff
@@ -155,6 +165,24 @@ public function getFallbackLocales(): array
         return $this->fallbackLocales;
     }
 
+    public function addGlobalParameter(string $id, mixed $value): void
```
this is not consistent with the property type that says `string|int|float`

stof commented on src/Symfony/Component/Translation/Translator.php:
```diff
@@ -174,6 +202,13 @@ public function trans(?string $id, array $parameters = [], ?string $domain = nul
         }
 
+        if ($this->globalTranslatableParameters && !isset($this->globalParameters[$locale])) {
+            $this->globalParameters[$locale] = []; // Avoid infinite loops
```
the storage of resolved global parameters per locale should be separate from the storage of the user-provided configuration IMO, instead of using a `*` key.

Jean-Beru replied:
Done using a `globalTranslatedParameters`.

stof commented on src/Symfony/Component/Translation/Translator.php:
```diff
@@ -174,6 +202,13 @@ public function trans(?string $id, array $parameters = [], ?string $domain = nul
 
+        if ($this->globalTranslatableParameters && !isset($this->globalParameters[$locale])) {
+            $this->globalParameters[$locale] = []; // Avoid infinite loops
+            $this->globalParameters[$locale] = array_map(fn ($parameter) => $parameter->trans($this, $locale), $this->globalTranslatableParameters);
```
Given that `$parameters` below has support for TranslatableInterface, do we actually need to handle those global parameters early ?

Jean-Beru replied:
We need to do this in order to avoid infinite loops since all parameters are translated next using the same `trans` method.

stof commented on src/Symfony/Component/Translation/Tests/TranslatorTest.php:
```diff
@@ -602,6 +600,56 @@ public function testMissingLoaderForResourceError()
+    }
+
+    /**
+     * @requires extension intl
```
Does this really requires the extension ? We support the ICU format without the extension thanks to our polyfill package.

Jean-Beru replied:
If we can use the polyfill, we may remove this requirement from all tests, right?

https://github.com/symfony/symfony/blob/6f4f04b98f6134996a57d21d1851d558a34dc45c/src/Symfony/Component/Translation/Tests/TranslatorTest.php#L561

stof commented on src/Symfony/Component/Translation/Tests/TranslatorTest.php:
```diff
@@ -399,14 +399,12 @@ public function testTrans($expected, $id, $translation, $parameters, $locale, $d
     }
 
     /**
+     * @requires extension intl
```
This is not the equivalent of the existing check

nicolas-grekas reviewed (commented):
<details>
<summary>Here is my review as a patch.</summary>

```diff
diff --git a/src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php b/src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php
index 44e9816e18..eba4ee4534 100644
--- a/src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php
+++ b/src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php
@@ -180,6 +180,7 @@ use Symfony\Component\Translation\Command\XliffLintCommand as BaseXliffLintComma
 use Symfony\Component\Translation\Extractor\PhpAstExtractor;
 use Symfony\Component\Translation\LocaleSwitcher;
 use Symfony\Component\Translation\PseudoLocalizationTranslator;
+use Symfony\Component\Translation\TranslatableMessage;
 use Symfony\Component\Translation\Translator;
 use Symfony\Component\TypeInfo\Type;
 use Symfony\Component\TypeInfo\TypeResolver\PhpDocAwareReflectionTypeResolver;
@@ -1612,11 +1613,7 @@ class FrameworkExtension extends Extension
         }
 
         foreach ($config['globals'] as $name => $global) {
-            if (isset($global['value'])) {
-                $translator->addMethodCall('addGlobalParameter', [$name, $global['value']]);
-            } else {
-                $translator->addMethodCall('addGlobalTranslatableParameter', [$name, $global['message'], $global['parameters'] ?? [], $global['domain'] ?? null]);
-            }
+            $translator->addMethodCall('addGlobalParameter', [$name, $global['value'] ?? new Definition(TranslatableMessage::class, [$global['message'], $global['parameters'] ?? [], $global['domain'] ?? null])]);
         }
 
         if ($config['pseudo_localization']['enabled']) {
diff --git a/src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/translation.html.twig b/src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/translation.html.twig
index 5babc0d893..18319e4993 100644
--- a/src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/translation.html.twig
+++ b/src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/translation.html.twig
@@ -172,18 +172,9 @@
                 <tr>
                     <td class="font-normal text-small nowrap"></td>
                     <td class="font-normal text-small nowrap">{{ id }}</td>
-                    <td class="font-normal text-small nowrap">{{ value }}</td>
+                    <td class="font-normal text-small nowrap">{{ profiler_dump(value) }}</td>
                 </tr>
             {% endfor %}
-            {% for locale, values in collector.globalTranslatedParameters %}
-                {% for id, value in values %}
-                    <tr>
-                        <td class="font-normal text-small nowrap">{{ locale }}</td>
-                        <td class="font-normal text-small nowrap">{{ id }}</td>
-                        <td class="font-normal text-small nowrap">{{ value }}</td>
-                    </tr>
-                {% endfor %}
-            {% endfor %}
             </tbody>
         </table>
     {% endif %}
diff --git a/src/Symfony/Component/Translation/CHANGELOG.md b/src/Symfony/Component/Translation/CHANGELOG.md
index 64502a89fd..365c5cf1cd 100644
--- a/src/Symfony/Component/Translation/CHANGELOG.md
+++ b/src/Symfony/Component/Translation/CHANGELOG.md
@@ -4,7 +4,7 @@ CHANGELOG
 7.3
 ---
 
- * Add `addGlobalParameter` and `addGlobalTranslatableParameter` to allow defining global translation parameters
+ * Add `Translator::addGlobalParameter()` to allow defining global translation parameters
 
 7.2
 ---
diff --git a/src/Symfony/Component/Translation/DataCollector/TranslationDataCollector.php b/src/Symfony/Component/Translation/DataCollector/TranslationDataCollector.php
index bb7dc6b399..e2e597a705 100644
--- a/src/Symfony/Component/Translation/DataCollector/TranslationDataCollector.php
+++ b/src/Symfony/Component/Translation/DataCollector/TranslationDataCollector.php
@@ -45,7 +45,6 @@ class TranslationDataCollector extends DataCollector implements LateDataCollecto
         $this->data['locale'] = $this->translator->getLocale();
         $this->data['fallback_locales'] = $this->translator->getFallbackLocales();
         $this->data['global_parameters'] = $this->translator->getGlobalParameters();
-        $this->data['global_translated_parameters'] = $this->translator->getGlobalTranslatedParameters();
     }
 
     public function reset(): void
@@ -94,14 +93,6 @@ class TranslationDataCollector extends DataCollector implements LateDataCollecto
         return $this->data['global_parameters'] ?? [];
     }
 
-    /**
-     * @internal
-     */
-    public function getGlobalTranslatedParameters(): Data|array
-    {
-        return $this->data['global_translated_parameters'] ?? [];
-    }
-
     public function getName(): string
     {
         return 'translation';
diff --git a/src/Symfony/Component/Translation/Translator.php b/src/Symfony/Component/Translation/Translator.php
index 6994cb9b6f..d57ec9e2c9 100644
--- a/src/Symfony/Component/Translation/Translator.php
+++ b/src/Symfony/Component/Translation/Translator.php
@@ -61,15 +61,10 @@ class Translator implements TranslatorInterface, TranslatorBagInterface, LocaleA
     private bool $hasIntlFormatter;
 
     /**
-     * @var array<string, string|int|float>
+     * @var array<string, string|int|float|TranslatableInterface>
      */
     private array $globalParameters = [];
 
-    /**
-     * @var array<string, TranslatableInterface>
-     */
-    private array $globalTranslatableParameters = [];
-
     /**
      * @var array<string, string|int|float>
      */
@@ -170,16 +165,11 @@ class Translator implements TranslatorInterface, TranslatorBagInterface, LocaleA
         return $this->fallbackLocales;
     }
 
-    public function addGlobalParameter(string $id, string|int|float $value): void
+    public function addGlobalParameter(string $id, string|int|float|TranslatableInterface $value): void
     {
         $this->globalParameters[$id] = $value;
     }
 
-    public function addGlobalTranslatableParameter(string $id, string $message, array $parameters = [], ?string $domain = null): void
-    {
-        $this->globalTranslatableParameters[$id] = new TranslatableMessage($message, $parameters, $domain);
-    }
-
     /**
      * @internal
      */
@@ -188,14 +178,6 @@ class Translator implements TranslatorInterface, TranslatorBagInterface, LocaleA
         return $this->globalParameters;
     }
 
-    /**
-     * @internal
-     */
-    public function getGlobalTranslatedParameters(): array
-    {
-        return $this->globalTranslatedParameters;
-    }
-
     public function trans(?string $id, array $parameters = [], ?string $domain = null, ?string $locale = null): string
     {
         if (null === $id || '' === $id) {
@@ -215,17 +197,24 @@ class Translator implements TranslatorInterface, TranslatorBagInterface, LocaleA
             }
         }
 
-        if ($this->globalTranslatableParameters && !isset($this->globalTranslatedParameters[$locale])) {
-            $this->globalTranslatedParameters[$locale] = []; // Avoid infinite loops
-            $this->globalTranslatedParameters[$locale] = array_map(fn ($parameter) => $parameter->trans($this, $locale), $this->globalTranslatableParameters);
+        foreach ($parameters as $key => $value) {
+            if ($value instanceof TranslatableInterface) {
+                $parameters[$key] = $value->trans($this, $locale);
+            }
         }
-        if ($this->globalParameters) {
-            $parameters += $this->globalParameters;
+
+        if (null === $globalParameters =& $this->globalTranslatedParameters[$locale]) {
+            $globalParameters = $this->globalParameters;
+            foreach ($globalParameters as $key => $value) {
+                if ($value instanceof TranslatableInterface) {
+                    $globalParameters[$key] = $value->trans($this, $locale);
+                }
+            }
         }
-        if (isset($this->globalTranslatedParameters[$locale])) {
-            $parameters += $this->globalTranslatedParameters[$locale];
+
+        if ($globalParameters) {
+            $parameters += $globalParameters;
         }
-        $parameters = array_map(fn ($parameter) => $parameter instanceof TranslatableInterface ? $parameter->trans($this, $locale) : $parameter, $parameters);
 
         $len = \strlen(MessageCatalogue::INTL_DOMAIN_SUFFIX);
         if ($this->hasIntlFormatter
```

</details>

The TL;DR is that I don't think that we need nor should have two methods for this.
Also, the PR is missing some cross-components checks I think (fwb / profiler are using feats that might not be available depending on the version of translator that's installed.)

Jean-Beru commented:
> Here is my review as a patch.
> The TL;DR is that I don't think that we need nor should have two methods for this. Also, the PR is missing some cross-components checks I think (fwb / profiler are using feats that might not be available depending on the version of translator that's installed.)

Thanks for the review :heart: 

> the PR is missing some cross-components checks

I updated the profiler part view but what about the bundle? Should we use `InstalledVersions::satisfies` to check component version, use a dedicated interface or just check the method existence?

stof commented:
The cleanest solution for FrameworkBundle is to bump the min supported version of `symfony/translation` (note that the `conflict` rule is the most important one here, as `require-dev` only impacts our CI)

Jean-Beru commented:
> The cleanest solution for FrameworkBundle is to bump the min supported version of `symfony/translation` (note that the `conflict` rule is the most important one here, as `require-dev` only impacts our CI)

Thanks for the tip!

xabbuh commented on src/Symfony/Component/Translation/Tests/TranslatorCacheTest.php:
```diff
@@ -55,14 +55,12 @@ protected function deleteTmpDir()
     }
 
     /**
+     * @requires extension intl
```
Is there a reason for this change? If I don't miss anything the test was runnable without the extension if the `symfony/polyfill-intl-messageformatter` was installed. This doesn't seem to be the case anymore.

Jean-Beru replied:
Honestly, I don't remember why I changed it. I revert it.

xabbuh commented on src/Symfony/Component/Translation/Translator.php:
```diff
@@ -174,7 +197,24 @@ public function trans(?string $id, array $parameters = [], ?string $domain = nul
         }
 
-        $parameters = array_map(fn ($parameter) => $parameter instanceof TranslatableInterface ? $parameter->trans($this, $locale) : $parameter, $parameters);
+        foreach ($parameters as $key => $value) {
```
Do we need this loop?

nicolas-grekas replied:
it's more efficient code since we *maybe* mutate an existing array instead of always copying it

xabbuh commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/translation.html.twig:
```diff
@@ -156,6 +156,27 @@
         {% endblock messages %}
     {% endif %}
 
+    {% if collector.globalParameters is defined and collector.globalParameters is not empty %}
```
can be shortened to

```suggestion
    {% if collector.globalParameters|default([]) %}
```

Jean-Beru replied:
Good catch.

xabbuh commented on src/Symfony/Component/Translation/Translator.php:
```diff
@@ -174,7 +197,24 @@ public function trans(?string $id, array $parameters = [], ?string $domain = nul
+            }
+        }
+
+        if (null === $globalParameters =& $this->globalTranslatedParameters[$locale]) {
```
What is the purpose of the `$globalTranslatedParameters` property? Is it there to cache already translated values? If so, isn't this broken for the case where the is a sequence of `trans()`, `addGlobalParameter()`, `trans()` calls?

nicolas-grekas replied:
Which case do you have in mind? IIUC, the cached data doesn't depend on the input so when would this break?

xabbuh replied:
After the first `trans` call `$globalParameters` isn't `null` anymore. So if you add new globals later on they wouldn't be translated on the next `trans()` call, right?

nicolas-grekas replied:
ah so we should empty this->globalTranslatedParameters when we add a global indeed!
in practice this shouldn't happen but better be safe

xabbuh replied:
And we should probably also implement the `ResetInterface()`?

nicolas-grekas replied:
Not sure about ResetInterface: are there any other factors to consider that'd need a reset?

Jean-Beru replied:
> If so, isn't this broken for the case where the is a sequence of trans(), addGlobalParameter(), trans() calls?

Indeed ! `$globalTranslatedParameters` is now emptied when using `addGlobalParameter`.

nicolas-grekas commented on src/Symfony/Component/Translation/DataCollectorTranslator.php:
```diff
@@ -82,6 +82,15 @@ public function getFallbackLocales(): array
 
+    public function getGlobalParameters(): array
+    {
+        if ($this->translator instanceof Translator || method_exists($this->translator, 'getGlobalParameters')) {
```
is the method_exists check needed?

xabbuh replied:
One could ship their own `TranslatorInterface` implementation having global parameters support too. But maybe having a dedicated interface might be a better solution.

nicolas-grekas replied:
this is `instanceof Translator` so the method is going to be always here to me, no?

xabbuh replied:
but `$this->translator` doesn't necessarily have to be an instance of `Translator` 🤔

nicolas-grekas replied:
Oh, NVM, I read the `||` as `&&` :facepalm:

nicolas-grekas commented on src/Symfony/Component/Translation/Translator.php:
```diff
@@ -155,6 +165,19 @@ public function getFallbackLocales(): array
+    }
+
+    /**
+     * @internal
```
let's remove this annotation? I don't see it helping us (maintainers) much, while it might be useful for userland, next to addGlobalParameter

Jean-Beru replied:
Sure ! It's done.

Is there any reason to keep it for the `getFallbackLocales` method ?

fabpot commented:
Thank you @Jean-Beru.
